migration/migration.go

```diff
@@ -300,3 +304,31 @@ func (m *changeTrieNodeEncoding) Migrate(txn db.Transaction, _ utils.Network) er
 	}
 	return iterator.Close()
 }
+
+// calculateBlockCommitments calculates the txn and event commitments for each block and stores them separately
+func calculateBlockCommitments(txn db.Transaction, network utils.Network) error {
+	var txnLock sync.RWMutex
+	workerPool := pool.New().WithErrors().WithMaxGoroutines(runtime.GOMAXPROCS(0))
+
+	for blockNumber := 0; ; blockNumber++ {
+		txnLock.RLock()
+		block, err := blockchain.BlockByNumber(txn, uint64(blockNumber))
+		txnLock.RUnlock()
```

Comment:
Are you using the Mutex to protect `block` or `blockNumber`? 

Also it looks like to me that you are using `BlockByNumber` only to determine when db no longer have blocks, if that is the case then I think the following code would be more simpler (by removing the need for mutex):

```go
// calculateBlockCommitments calculates the txn and event commitments for each block and stores them separately
func calculateBlockCommitments(txn db.Transaction, network utils.Network) error {
	workerPool := pool.New().WithErrors().WithMaxGoroutines(runtime.GOMAXPROCS(0))

	for blockNumber := uint64(0); ; blockNumber++ {
		block, err := blockchain.BlockByNumber(txn, blockNumber)

		if errors.Is(err, db.ErrKeyNotFound) {
			break
		}

		workerPool.Go(func() error {
			block := block
			commitments, err := core.VerifyBlockHash(block, network)
			if err != nil {
				return err
			}
			return blockchain.StoreBlockCommitments(txn, block.Number, commitments)
		})
	}

	return workerPool.Wait()
}
```

Reply:
As the name implies, mutex is used to protect the txn. We are concurrently doing both read and write operations on it. So it needs to be protected.

Reply:
I don't understand why we need to protect the `txn`, should pebble handle concurrent updates and read to batches/snapshot also we have a test for ConcurrentUpdates. 

Also, `block := block` is required

Reply:
> Also, block := block is required

I dont think so, the variable is in the `for` loop body. So they are separate instances.

> I don't understand why we need to protect the txn, should pebble handle concurrent updates and read to batches/snapshot also we have a test for ConcurrentUpdates.

Transactions are not concurrency-safe. What `TestConcurrentUpdate` is testing is that db can be read and updated concurrently concurrently thru `db.Update`/`db.View`, which each create separate transactions.